src/pybind/mgr/deepsea/module.py

```diff
@@ -0,0 +1,290 @@
+        event = {}
+        try:
+            for line in self._event_response.iter_lines():
+                if line:
+                    line = line.decode('utf-8')
+                    colon = line.find(':')
+                    if colon > 0:
+                        k = line[:colon]
+                        v = line[colon+2:]
+                        if k == "retry":
+                            # TODO: find out if we need to obey this reconnection time
+                            self.log.warn("Server requested retry {}, ignored".format(v))
+                        else:
+                            event[k] = v
+                else:
+                    # Empty line, terminates an event.  Note that event['tag']
+                    # is a salt-api extension to SSE to avoid having to decode
+                    # json data if you don't care about it.  To get to the
+                    # interesting stuff, you want event['data'], which is json.
+                    # TODO: this should probably really be debug level
+                    self.log.info("Got event '{}'".format(str(event)))
+
+                    # If we actually wanted to do something with the event,
+                    # say, we want to notice that some long salt run has
+                    # finished, we'd call some notify method here (TBD).
+
+                    # If you want to have some fun, try
+                    # `ceph daemon mgr.$(hostname) config set debug_mgr 4/5`
+                    # then `salt '*' test.ping` on the master
+                    event = {}
+            self._set_last_failure_msg("SSE read terminated")
+        except Exception as ex:
+            self._set_last_failure_msg("SSE read failed: {}".format(str(ex)))
+
+        self._reading_events = False
+        self.event.set()
+
+
+    # _do_request(), _login() and _do_request_with_login() are an extremely
+    # minimalist form of the following, with notably terse error handling:
+    # https://bitbucket.org/openattic/openattic/src/ce4543d4cbedadc21b484a098102a16efec234f9/backend/rest_client.py?at=master&fileviewer=file-view-default
+    # https://bitbucket.org/openattic/openattic/src/ce4543d4cbedadc21b484a098102a16efec234f9/backend/deepsea.py?at=master&fileviewer=file-view-default
+    # rationale:
+    # - I needed slightly different behaviour than in openATTIC (I want the
+    #   caller to read the response, to allow streaming the salt-api event bus)
+    # - I didn't want to pull in 400+ lines more code into this presently
+    #   experimental module, to save everyone having to review it ;-)
+
+    def _do_request(self, method, path="", data=None, stream=False):
+        """
+        returns the response, which the caller then has to read
+        """
+        protocol = 'https' if self.get_config('salt_api_ssl').lower() != 'false' else 'http'
+        url = "{0}://{1}:{2}/{3}".format(protocol,
+                                         self.get_config('salt_api_host'),
```

Comment:
Is this IPv6 ready? In the dashboard, we created [`build_url()`](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/tools.py#L668) to handle that case.

Reply:
No, it's not. Good catch!

Reply:
Hrm. So, I can either copy `build_url()` and `is_valid_ipv6_address()` into this module (yuck), or, is there somewhere we can put generic utility stuff like this so all modules can use those functions? @jcsp?

Reply:
I'm still wondering, why Python doesn't provide a proper way to handle URLs. I'd be +1 for adding a Python module containing miscellaneous utilities usable by all modules.

Reply:
...or, I could just merge the host, port and ssl options into a single URL option (approximately as @jcsp suggested for #24445) and the whole problem goes away :-)

Reply:
Done